e2e loadbalancer udp conntrack test

aojea opened the pull request:
The test verifies that a LoadBalancer UDP service doesn't blackhole the traffic to the node when the pod backend is destroyed and the traffic has to fall back to another pod.


/kind cleanup

```release-note
NONE
```

Tested against a gke cluster with dpv2

```
 _output/local/bin/linux/amd64/e2e.test  --ginkgo.focus="should be able to preserve UDP traffic when server pod cycles for a LoadBalancer service" --context  gke_aojea-gke-dev_us-central1-c_cluster-2  --kubeconfig $HOME/.kube/config --provider gke --gce-zone us-central1

...


Ran 1 of 6610 Specs in 71.272 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 6609 Skipped
PASS

```

aojea commented:
/assign @MrHohn @panslava @code-elinka 
/cc @bowei @czahedi  @swetharepakula

aojea committed:
e2e loadbalancer udp conntrack test

aojea commented:
/assign @cezarygerard

thockin reviewed (commented):
This test looks OK, but why is it only GCE?

aojea commented on test/e2e/network/loadbalancer.go:
```diff
@@ -969,6 +972,138 @@ var _ = common.SIGDescribe("LoadBalancers", func() {
+
+	ginkgo.It("should be able to preserve UDP traffic when server pod cycles for a LoadBalancer service", func() {
+		// requires cloud load-balancer support
+		e2eskipper.SkipUnlessProviderIs("gce", "gke")
```
I copied from the others , honestly, I don't know the origin of these tests and I'm afraid of breaking someone because suddenly some tests start to run

aojea replied:
updated to match all cloud providers

aojea commented:
> This test looks OK, but why is it only GCE?

🤷  culto cargoing with the current state

aojea commented:
should I change them?

There is one with 

> 		e2eskipper.SkipUnlessProviderIs("gce", "gke", "aws")

cezarygerard commented on test/e2e/network/loadbalancer.go:
```diff
@@ -969,6 +972,138 @@ var _ = common.SIGDescribe("LoadBalancers", func() {
+		nodes, err := e2enode.GetBoundedReadySchedulableNodes(cs, 2)
+		framework.ExpectNoError(err)
+		if len(nodes.Items) < 2 {
+			e2eskipper.Skipf(
```
Skip, not fail?

aojea replied:
we don't have a good criteria here, I think that we can justify any action ehre, but I prefer to be gentle and not make the test fail to avoid surprises on consumers of the tests

cezarygerard commented:
Antonio,

Maybe it's a silly question,
but I don't understand why you use 2 nodes for your 2 pods in this test.
FWIU UDP contract timeout can only blackhole traffic if pod disappeared and the traffic was send to the same node.

If the only available pod is on another node, then packet should be routed normally, and there is no existing UDP connection on the other node, so everything will always work, regardless if kubernetes cleans the contrack table after pod deletion or not.


Also I feel this test has nothing GCE/GKE specific

aojea commented:
> If the only available pod is on another node, then packet should be routed normally, and there is no existing UDP connection on the other node, so everything will always work, regardless if kubernetes cleans the contrack table after pod deletion or not.

good question, I was assuming the LB keeps sending the traffic to the same node because is the same session (same ip,port tuple= and is the kube-proxy on that node the one forwarding the traffic , is that wrong?

anyway, I will add another case with the pods on the same node, so we can verify both cases,

aojea commented:
addressed the comments and added a new test case for handling both scenarios
1. pods spawned on different nodes
2. pods spawned on same node

https://gist.github.com/aojea/cd5bdd6052b509280e0996962bfbd4a9

dims commented:
/assign @thockin 

just 2 days to test freeze!

cezarygerard commented:
/lgtm

aojea commented:
/milestone v1.26

increasing test coverage before test freeze
